lib/spack/spack/build_systems/intel.py

```diff
@@ -1241,6 +1241,32 @@ def configure_rpath(self):
             with open(compiler_cfg, 'w') as fh:
                 fh.write('-Xlinker -rpath={0}\n'.format(compilers_lib_dir))
 
+    @run_after('install')
+    def configure_auto_dispatch(self):
+        if self._has_compilers:
+            if ('auto_dispatch=none' in self.spec):
+                return
+
+            compilers_bin_dir = self.component_bin_dir('compiler')
+
+            for compiler_name in 'icc icpc ifort'.split():
+                f = os.path.join(compilers_bin_dir, compiler_name)
+                if not os.path.isfile(f):
+                    raise InstallError(
+                        'Cannot find compiler command to configure '
+                        'auto_dispatch:\n\t' + f)
+
+                ad = []
+                for x in ('COMMON-AVX512', 'MIC-AVX512', 'CORE-AVX512',
```

Comment:
This set of strings is repeatedly defined in separate packages. Could you define a single tuple for this in this module and reuse it in the other two `IntelPackage`s?

An object might actually be better because then you can define fields like `.SSE2` that could be used in the conflict declarations (to make sure that there aren't typos when folks edit things in the future).

Reply:
@scheibelp 
That is a minor request but I can not get the package files to read the tuple from the IntelPackage class. Are there any examples of something similar elsewhere in spack? I am probably missing something simple.

Reply:
I figured out a way to get the variable passed. But using that in the variant does not seem to work.
I added a couple of print statements for debugging.
```
$ spack spec intel auto_dispatch=CORE-AVX2
Input spec
--------------------------------
intel auto_dispatch=CORE-AVX2

Concretized
--------------------------------
<type 'tuple'>
Values are: ('COMMON-AVX512', 'MIC-AVX512', 'CORE-AVX512', 'CORE-AVX2', 'CORE-AVX-I', 'AVX', 'SSE4.2', 'SSE4.1', 'SSSE3', 'SSE3', 'SSE2')
==> Error: invalid values for variant "auto_dispatch" in package "intel": ['CORE-AVX2']
```

Looking at `spack info`, the variant options do not look correct, as the parenthesis are in the list.
```
spack info
...
Variants:
    Name [Default]          Allowed values          Description


    auto_dispatch [none]    none, ('COMMON-         Enable generation of multiple
                            AVX512', 'MIC-          auto-dispatch code paths
                            AVX512', 'CORE-
                            AVX512', 'CORE-
                            AVX2', 'CORE-AVX-I',
                            'AVX', 'SSE4.2',
                            'SSE4.1', 'SSSE3',
                            'SSE3', 'SSE2')
    rpath [on]              True, False             Add rpath to .cfg files
```

Reply:
It's hard for me to tell 100% without seeing the changes you made, but it looks like you called `any_combination_of(the_tuple)` when instead you'll need to call `any_combination_of(*the_tuple)` (i.e. you need to use the `*` syntax to expand the sequence into a set of arguments to the function). If that's not what you did, could you push the changes you made?

Reply:
@scheibelp 
Sorry I did not paste code in my comment. You are exactly right, I was missing the '*'. I will be pushing more changes in a bit that has things working. I was struggling with the inheritance a bit so am not sure I have done it the best way.

Reply:
As I was going through this it occurred to me that having that variant and the corresponding conflict statements would be good to have in `IntelPackage`. I was not sure how to get the package name at the Parent class level though. Without a check for the package name the variant would be present for all Intel packages, but of course only makes sense for the compiler packages.